### PR TITLE
[NL] Apply all Dirichlet BCs at once.

### PR DESCRIPTION
**Pull request by TomFischer**
Saves time.

### CODE REVIEW

**Comment by endJunction** [comment]
Looks good. Could you post some timings, like for the "Unstrut part" master takes ~x secs... (Only what you already have.)
Would you try to sort the ids (simultaneously with the values) and measure the times one more time please?

File: NumLib/ODESolver/TimeDiscretizedODESystem.h [hunk 1]
```diff
@@ -359,10 +359,13 @@ class TimeDiscretizedODESystem<Matrix, Vector,
             _ode.getKnownSolutions(_time_disc.getCurrentTime());
 
         if (known_solutions) {
+            decltype((*known_solutions)[0].ids) ids;
```
Comment (chleh) [inline NumLib/ODESolver/TimeDiscretizedODESystem.h]: That looks very awkward.

**Comment by chleh** [comment]
A different option is to extend `MathLib::applyKnownSolution` s.t. it accepts the whole vector `known_solutions` and loop over it there.

**Comment by TomFischer** [comment]
As @endJunction suggested I will try the sort the ids and values and will produce timings.

**Comment by TomFischer** [comment]
I did some time measurements.
Without the commit 6de4fb2 `applyKnownSolution`  took 37078 seconds (10:20 h). The reduce the time to circa 8 seconds. The sorting and application of the id and value vector the time for `applyKnownSolution` is 12 seconds.

**Comment by TomFischer** [comment]
> A different option is to extend MathLib::applyKnownSolution s.t. it accepts the whole vector known_solutions and loop over it there.

The linear algebra functionality deals with matrices and vectors. I don't like the idea to offer an interface that accepts boundary condition object in the linear algebra data structures. As long as this is not a bottleneck anymore I would keep the implementation as simple as possible.

**Comment by endJunction** [comment]
:fast_forward: 
I'd replace the `decltype` things with explicit types, which are known and fixed. The `decltype` is usually used in template metaprogramming where the types can be different for the same algorithm...

**Comment by chleh** [comment]
:+1:

**Comment by wenqing** [comment]
:+1:  It is also better for MPI stuff.

**Pull request merged**